src/main/java/com/github/novelrt/fumocement/NativeObject.java

```diff
@@ -22,49 +31,122 @@
+   * {@code owned} parameter is {@code false}.
+   *
+   * @param handle        the native handle
+   * @param owned         whether or not this object owns native resources
+   * @param handleDeleter the {@link HandleDeleter} to use in order to delete native resources,
+   *                      which can be {@code null} when {@code owned} is false
+   */
+  protected NativeObject(@Pointer long handle, boolean owned, HandleDeleter handleDeleter) {
+    this.state = new State(handle, handleDeleter, owned ? ResourceOwningState.OWNED : ResourceOwningState.UNOWNED);
+    if (owned) {
+      if (handleDeleter == null) {
+        throw new IllegalArgumentException("This NativeObject is owned, but its handleDeleter is null.");
+      }
       NATIVE_OBJECTS_CLEANER.register(this, state);
     }
   }
 
-  protected long getHandle() {
+  /**
+   * Gets the native handle.
+   * <p>
+   * This handle is represented as a {@code long} to support 64-bit pointers.
+   *
+   * @throws IllegalStateException when the native resource this object holds has been cleared
+   * @return the native handle
+   */
+  protected @Pointer long getHandle() {
+    if (state.resourceOwningState == ResourceOwningState.CLEARED) {
+      throw new IllegalStateException("Cannot get the handle of this NativeObject as its native resource has " +
+                                      "already been cleared.");
+    }
     return state.handle;
   }
 
-  protected boolean isOwned() {
-    return isOwned;
+  /**
+   * Gets the the current state of native resources owned by this object.
+   * <p>
+   * This determines if there needs to be a native resource releasing process
+   * (a {@link HandleDeleter}) once the object is garbage collected.
+   *
+   * @return a {@link ResourceOwningState} indicating the current state of native resources
+   *         owned by this object
+   */
+  protected ResourceOwningState getResourceOwningState() {
+    return state.resourceOwningState;
+  }
+
+  /**
+   * Clears any native resources this object holds.
+   * <p>
+   * This does not have any effect when the current resource owning state is not
+   * {@link ResourceOwningState#OWNED}.
+   */
+  @Override
+  public void close() {
+    state.deleteNativeResources();
+  }
+
+  /**
+   * Indicates the state for native resources contained in a {@link NativeObject}.
+   */
+  public enum ResourceOwningState {
+    /**
+     * The native resource is not managed by the JVM, the object only serves as a access layer
+     * to the resource, without disposing anything.
+     */
+    UNOWNED,
+    /**
+     * The native resource is managed by the JVM. It can be cleaned manually as well as it can be
+     * cleaned once the object has been garbage collected.
+     */
+    OWNED,
+    /**
+     * The native resource has definitively been removed. Any call to {@link NativeObject#getHandle()}
+     * will produce an exception.
+     */
+    CLEARED
```

Comment:
Is this just Java convention?

Reply:
I'm not sure what you are talking about, is it SCREAMING_CASE_ENUMS? In this case yes, it's a java convention.